### PR TITLE
test: Handle Filter flake with ctx errors

### LINKED ISSUE #7114: RBAC filtering truncates filterset instead of returning error if context expires
When we use the RBAC filter, you pass a context.  If the context expires before we're done filtering, we return a truncated set of objects passing the filter and no error.  We _should_ return an empty filterset and whatever error came from the context.

This leads to strange test flakes like:

```
=== FAIL: coderd/rbac TestFilter/ManyRoles (5.07s)
    authz_internal_test.go:197: 
        	Error Trace:	/home/runner/work/coder/coder/coderd/rbac/authz_internal_test.go:197
        	Error:      	Not equal: 
        	            	expected: 32
        	            	actual  : 29
        	Test:       	TestFilter/ManyRoles
        	Messages:   	expected number of allowed

=== FAIL: coderd/rbac TestFilter (0.00s)
```

### PR DESCRIPTION
**Pull request by Emyrk**
Fixes: https://github.com/coder/coder/issues/7114

### CODE REVIEW

**Review by johnstcn — COMMENTED** [review]
This _looks_ right to me, I think this is a tricky flake to hit.
Can we add a test to assert this behaviour?

File: coderd/rbac/authz.go [hunk 1]
```diff
@@ -137,6 +137,9 @@ func Filter[O Objecter](ctx context.Context, auth Authorizer, subject Subject, a
 			err := auth.Authorize(ctx, subject, action, o.RBACObject())
 			if err == nil {
 				filtered = append(filtered, o)
+			} else if ctx.Err() != nil {
+				// Exit early if the error comes from the context
+				return nil, ctx.Err()
```
Comment (spikecurtis) [inline coderd/rbac/authz.go:142]: This code implicitly assumes that there are only 2 classes of errors that could be returned from `auth.Authorize`

1. the authorization completed, but the subject is not authorized
2. the context is Done

Prior to this change, this code implicitly assumed that there was only 1 class of errors.

It is difficult to be sure that the right answer is 2.  Could it be 3 classes?  4?

Honestly, I think a better design would be for `auth.Authorize` to return two values --- one that tells you whether the subject is authorized, and another to tell you whether or not there was an error checking the authorization.

Failing that, it's generally better for error checking to be explicit and specific:

```
err := auth.Authorize(...)
if xerrors.Is(err, rbac.Unauthorized) {
    continue
}
if err != nil {
    return nil, err
}
filtered = append(filtered, o)
```

**Review by spikecurtis — COMMENTED** [review]
A variation on this bug probably exists in all the places in the code we call `Authorize`

**Comment by spikecurtis** [comment]
> This _looks_ right to me, I think this is a tricky flake to hit. Can we add a test to assert this behaviour?

Yeah, a test would be good.  We generate a list of say, 3 `Objecter`s --- but the middle one is a little bomb that, as a side effect to it's `RBACObject()` call, cancels the context.  Assert that we get the error back instead of a list of 2.

File: coderd/rbac/authz.go (hunk 1 above)
Comment (spikecurtis) [reply coderd/rbac/authz.go:142]: Actually, thinking about this more, it's not about implicit vs explicit.  It's about how the code reacts if `Authorize()` returns an _unexpected_ error.  By only handling the context case, we'd be the in the same boat again of swallowing the unexpected error and treating it like an unauthorized response.

If we instead handle the _expected_ errors, and return any others, we're less likely to incorrectly ignore them.

Comment (Emyrk) [reply coderd/rbac/authz.go:142]: Here are the 2 classes of error returned by `authorize`:

https://github.com/coder/coder/blob/6d41edceae37ac6e0494d3951944674b8f881dda/coderd/rbac/authz.go#L326-L333

Right now both are `ForbiddenWithInternal`. Maybe the correct solution here is to return something different on the `Eval` error.

Comment (Emyrk) [reply coderd/rbac/authz.go:142]: Then I can check for `rbac.IsUnauthorizedError()` on the `Filter` function loop

Comment (Emyrk) [reply coderd/rbac/authz.go:142]: Actually it gets a bit more annoying in the prepared one.

https://github.com/coder/coder/blob/6d41edceae37ac6e0494d3951944674b8f881dda/coderd/rbac/authz.go#L434-L434

:thinking:

File: coderd/rbac/authz.go [hunk 2]
```diff
@@ -319,7 +327,8 @@ func (a RegoAuthorizer) authorize(ctx context.Context, subject Subject, action A
 
 	results, err := a.query.Eval(ctx, rego.EvalParsedInput(astV))
 	if err != nil {
-		return ForbiddenWithInternal(xerrors.Errorf("eval rego: %w", err), subject, action, object, results)
+		err = correctCancelError(err)
+		return xerrors.Errorf("evaluate rego: %w", err)
```
Comment (Emyrk) [inline coderd/rbac/authz.go:331]: For authorize single items.

File: coderd/rbac/authz.go [hunk 3]
```diff
@@ -430,7 +439,8 @@ EachQueryLoop:
 		// We need to eval each query with the newly known fields.
 		results, err := q.Eval(ctx, rego.EvalParsedInput(parsed))
 		if err != nil {
-			continue EachQueryLoop
+			err = correctCancelError(err)
+			return xerrors.Errorf("eval error: %w", err)
```
Comment (Emyrk) [inline coderd/rbac/authz.go:443]: For prepared.

**Review by Emyrk — COMMENTED** [review]
Unit test added.
Correctly made 2 categories of errors.

I also added some code to clean up `cancelled` errors to be more uniform with expected `context.Cancelled`

**Review by johnstcn — APPROVED** [review]

**Comment by spikecurtis** [comment]
> Ugh a unit test is really hard to make for this.
> 
> The ctx is checked for cancel in a go routine: https://github.com/open-policy-agent/opa/blob/main/rego/rego.go#L2002-L2004
> 
> If the policy executes fast enough, then the ctx is ignored. Which means this is racy.

The standard procedure for when your code needs to call some 3rd party thing that is non-deterministic, is to mock it out and have a test for each case/class of what the non-deterministic thing can return.

**Pull request merged**